nsv13/code/_onclick/drag_drop.dm

```diff
@@ -0,0 +1,74 @@
+/obj/item/gun
+	var/atom/autofire_target = null //What are we aiming at? This will change if you move your mouse whilst spraying.
+
+/obj/item/gun/Initialize()
+	. = ..()
+	canMouseDown = automatic
+
+/obj/item/gun/onMouseUp(object, location, params, mob)
+	autofire_target = null
+	return ..()
+
+/obj/item/gun/proc/can_fire_at(atom/target, mob/user)
+	if(!target)
+		return FALSE
+
+	if(user.get_active_held_item() != src)
+		return FALSE
+
+	if(istype(target, /obj/screen))
+		return FALSE
+
+	if(target in user.contents) //can't shoot stuff inside us.
+		return FALSE
+
+	if(user.a_intent == INTENT_HARM) //melee attacks are handled by attackby, not autofire
+		return FALSE
+
+	if(target == user) //so we can't shoot ourselves with autofire
+		return FALSE
+
+	if(istype(user))//Check if the user can use the gun, if the user isn't alive(turrets) assume it can.
+		var/mob/living/L = user
+		if(!can_trigger_gun(L))
+			return FALSE
+
+	if(!can_shoot()) //Just because you can pull the trigger doesn't mean it can shoot. (*CLICK*)
+		reset_autofire()
+		return FALSE
+
+	if(clumsy_check)//If theyre a clown, they can't autofire.
+		if(istype(user))
+			if (HAS_TRAIT(user, TRAIT_CLUMSY))
```

Comment:
I feel like this should give the user feedback.

Reply:
It does, actually. But just on the first click, rather than spamming them with it twice.